---
title: "Wednesday May 19th Meeting"
author: "Gheeda Mourtada"
date: "19/05/2021"
output: beamer_presentation

---

Running the _SpatialExperiment_ vingette


\newpage

## Introduction

Firstly, it's mentioned that the package provides an S4 class for Spatial Omics data. This is just a software in R that allows creating objects in a stricter sense where they have **a formal definition** and a regular way to be created and are open to validaton.

The `SpatialExperiment` package is available via Bioconductor. It inherits from the
`SingleCellExperiment` class, which also inherits from the `SummarizedExperiment` and is used in the same manner. 

To install the package, we can use the [bioconductor](http://www.bioconductor.org/packages/release/bioc/html/SpatialExperiment.html) website. 

```{r include=FALSE}
#loading the package
library(SpatialExperiment)
```

Next, depending on the type of data used, we can load our data inside the SpatialExperiment class.

----

## SpatialExperiment has the following components

- **assays:** genes cells count matrix
- **rowData:** gene information
- **colData:** cell/sample information
- **spatialCoords:** spatial information _(including sequencing depth)_

----

![Illustration of SpatialExperiment Components](A-Illustration-of-the-structure-of-the-SpatialExperiment-class-including.png){fig.align='center'}

----

### Defining seqFISH 

Abbreviation of sequential Fluorescence _In Situ_ Hybridization)

**Goal:** identifying many RNA transcripts in single cells while maintaining spatial information 

**Method:** repeatedly performing sequential rounds of hybridization with fluorescent probes that result in a temporal barcode. This is what allows the RNA to be directly identified from samples.
**Note:** each transcript appears as a single dot and observed dots are quantified as abundance.
![Imported from seqfish.com](seqFISH.png)

----

### Simulated SeqFISH Data Workflow

Using the examples given on the _SpatialExperiment_ help page we simulate the following data:

\footnotesize
```{r}
# building random seqFISH data coordinates
fishCoordinates <- data.frame(Cell_ID=paste0("cell",c(1:30)),
                Irrelevant=100,
                x=sample(c(-4000:4000), size=30, replace=TRUE),
                y=sample(c(-4000:4000), size=30, replace=TRUE))

head(fishCoordinates)
```
\normalsize

----

\footnotesize
```{r}
# building random seqFISH cell labels
fishCellLabels <- data.frame(Cell_ID=paste0("cell",c(1:30)),
                             class="neuron",
                             classID=sample(c(0:5), size=30, replace=TRUE))

head(fishCellLabels)
```
\normalsize

----

\footnotesize
```{r}
# building random seqFISH count matrix
fishCounts <- matrix(sample(0:100, size=(30*30), replace=TRUE),
                     nrow=30, ncol=30,
                     dimnames=list(paste0("gene",c(1:30)),
                                   paste0("cell",c(1:30))))
head(fishCounts)
```
\normalsize

----

\footnotesize
```{r}
# creating SpatialExperiment object
se <- SpatialExperiment(rowData=rownames(fishCounts),
                        colData=fishCellLabels,
                        assays=SimpleList(counts=as.matrix(fishCounts)),
                        spatialCoords=fishCoordinates)
show(se)
```
\normalsize

----

First, to get the spatial coordinates, we use spatialCoords methods:

```{r}
spatialCoords(se)
```

----

Next, I set the spatial coordiantes, create a fake fish coordinates data frame overwrite the old coordinates data frame:

\footnotesize
```{r}
fakeFishCoords <- cbind(fishCoordinates[,c(1:3)], fishCoordinates[,3])
colnames(fakeFishCoords) <- c("Cell_ID", "Irrelevant", "x", "y")
spatialCoords(se) <- fakeFishCoords
spatialCoords(se)
```
\normalsize

----

### SeqFISH Data Workflow

I first loaded the provided seqFISH that's listed in the github data folder and set my directory to the folder that contains those files. The seqFISH data holds information about both the spatial coordinates and gene transcripts (while in contrast, single cell RNA sequencing has only gene transcipt information which is processed through `Single Cell Experiment`)

Retrieving and assigning variables to the data:
\footnotesize
```{r}
fishCoordFile <- system.file(file.path("extdata", "seqFISH",
                            "fcortex.coordinates.txt"), 
                            package="SpatialExperiment")
fishCoordinates <- read.table(fishCoordFile, header=FALSE, sep=" ")
colnames(fishCoordinates) <- c("Cell_ID", "Irrelevant", "x", "y")
fishCellLabsFile <- system.file(file.path("extdata", "seqFISH", 
                            "seqfish_cell_labels.tsv"),
                            package="SpatialExperiment")
fishCellLabels <- read.table(file=fishCellLabsFile, header=FALSE, sep="\t")
colnames(fishCellLabels) <- c("Cell_ID", "cluster", "class", "classID", 
                                "Irrelevant", "Prob")

fishFeatCountsFile <- system.file(file.path("extdata", "seqFISH",
                            "seqfish_normalized_cortex_b2_testing.txt"), 
                            package="SpatialExperiment")
fishFeaturesCounts <- read.table(file=fishFeatCountsFile, 
                                header=FALSE, sep="\t", row.names=1)

```
\normalsize

----

### Class handling

\footnotesize
```{r}
se <- SpatialExperiment(rowData=rownames(fishFeaturesCounts),
                        colData=fishCellLabels,
                        assays=SimpleList(counts=as.matrix(fishFeaturesCounts)),
                        spatialCoords=fishCoordinates)
show(se)
```
\normalsize

----

### Getter

```{r}
spatialCoords(se)
```

----

### Setter

\footnotesize
```{r}
fakeFishCoords <- cbind(fishCoordinates[,c(1:3)], fishCoordinates[,3])
colnames(fakeFishCoords) <- c("Cell_ID", "Irrelevant", "x", "y")
spatialCoords(se) <- fakeFishCoords
spatialCoords(se)
```
\normalsize

----

## Revisiting our objectives
**How can we use SpatialExperiment?**

- Building on other tools/implementing the method in different examples
- Reliably identifying unique tumor environments to be able to predict prognosis 
- Using the information we get from our predictions to have personalized treatments

**Refining our objectives:**

- Learn about the tools and how to integrate information
- Experiment with reproducible workflow
- Continue to refine the objective to get a clear goal

----

## References

[Beamer presentation](https://bookdown.org/yihui/rmarkdown/beamer-presentation.html)

[Bioconductor spatial transcriptomics talk](https://www.youtube.com/watch?v=bD_zaBpzCKY)

[Bioconductor SpatialExperiment workshop](https://www.youtube.com/watch?v=adCaPjAZ7xE&t=1067s)

[SeqFISH information](https://www.seqfish.com/technology)

[SpatialExperiment vignette](https://github.com/drighelli/SpatialExperiment)

Special thanks to Dr. Jeganathan and Kaiden for answering my questions! :)


